src/main/java/com/devonfw/ide/sonarqube/common/impl/check/DevonUcImplSecurityConstraintCheck.java

```diff
@@ -0,0 +1,159 @@
+ * security constraint from javax.annotation.security.
+ */
+@Rule(key = "E7", name = "Devon Uc Impl Security Constraint Check", //
+    priority = Priority.CRITICAL, tags = { "architecture-violation", "devonfw" })
+public class DevonUcImplSecurityConstraintCheck extends DevonArchitectureCheck {
+
+  private static final Set<String> REQUIRED_ANNOTATIONS = new HashSet<>(
+      Arrays.asList("DenyAll", "PermitAll", "RolesAllowed"));
+
+  /**
+   * Method called after parsing and semantic analysis has been done on file.
+   *
+   * @param context Context of analysis containing the parsed tree.
+   */
+  @Override
+  public void scanFile(JavaFileScannerContext context) {
+
+    ClassTree tree = getClassTree(context);
+    TypeTree ucInterface = getUcInterface(tree);
+
+    if (ucInterface == null) {
+      return;
+    }
+
+    List<MethodTree> methodsOfTree = getMethodsOfTree(tree);
+    for (MethodTree method : methodsOfTree) {
+      if (!isMethodProperlyAnnotated(method)) {
+        context.reportIssue(this, method,
+            "This method is not properly annotated. "
+                + "Please use one of the following annotations on Use-Case implementation methods: "
+                + REQUIRED_ANNOTATIONS.toString());
+      }
+    }
+  }
+
+  /**
+   * @param context Context of analysis containing the parsed tree.
+   * @return Class tree.
+   */
+  protected ClassTree getClassTree(JavaFileScannerContext context) {
+
+    CompilationUnitTree parsedTree = context.getTree();
+    List<Tree> types = parsedTree.types();
+
+    for (Tree tree : types) {
+      if (tree instanceof ClassTree) {
+        return (ClassTree) tree;
+      }
+    }
+
+    return null;
+  }
+
+  /**
+   * Returns the Uc interface implemented by the checked class, if there is one.
+   *
+   * @param tree Tree currently being investigated.
+   * @return TypeTree instance of the Uc interface.
+   */
+  protected TypeTree getUcInterface(ClassTree tree) {
+
+    List<TypeTree> interfaces = tree.superInterfaces();
+    String className = tree.simpleName().name();
+    Pattern interfaceRegEx = Pattern.compile(className.replaceAll("Impl", ""));
+
+    if (interfaces.isEmpty()) {
+      return null;
+    }
+
+    for (TypeTree interfaceTree : interfaces) {
+      if (interfaceRegEx.matcher(interfaceTree.toString()).matches()) {
+        return interfaceTree;
+      }
+    }
+
+    return null;
+  }
+
+  /**
+   * Returns all methods of the given tree.
+   *
+   * @param tree Tree currently being investigated.
+   * @return List of MethodTree.
+   */
+  protected List<MethodTree> getMethodsOfTree(ClassTree tree) {
+
+    List<Tree> membersOfTree = tree.members();
+    List<MethodTree> methodsOfTree = new ArrayList<>();
+
+    for (Tree member : membersOfTree) {
+      if (member.is(Tree.Kind.METHOD)) {
+        methodsOfTree.add((MethodTree) member);
+      }
+    }
+
+    return methodsOfTree;
+  }
```

Comment:
Nice utility method. As this could later be useful for other rules/checks, you could move this up to the super-class of all our checks.